File: Python/BOJ/스택의활용/4949.py
import sys

while True:
  input = sys.stdin.readline().rstrip()
  if input == ".":
    break
  
  arr = list(input)
  isError = False

  stack = []
  for char in arr:
    if char == "(" or char == "{" or char == "[":
      stack.append(char)
    elif char == ")" or char == "}" or char == "]":
      if not stack:
        isError = True
        break
        
      if char == ")":
        if stack[-1] == "(":
          stack.pop()
        else:
          isError = True
          break
      elif char == "}":
        if stack[-1] == "{":
          stack.pop()
        else:
          isError = True
          break
      elif char == "]":
        if stack[-1] == "[":
          stack.pop()
        else:
          isError = True
          break
      
  if isError or stack: print("no")
  else: print("yes")

File: Swift/BOJ/문제집/스택/4949.swift
//
//  4949.swift
//  Algorithm
//
//  Created by 유정주 on 2023/07/04.
//

import Foundation

let dict: [Character: Character] = [")": "(", "]": "["]

while let string = readLine(), string != "." {
    var stack: [Character] = []
    var isBalance = true
    
    let arr = string.filter { ["(", ")", "[", "]"].contains($0) }
    for char in arr {
        if char == "(" || char == "[" {
            stack.append(char)
        } else {
            guard let top = stack.last else {
                isBalance = false
                break
            }
            
            if let key = dict[char], key == top {
                stack.removeLast()
            } else {
                isBalance = false
                break
            }
        }
    }
    
    if !stack.isEmpty {
        isBalance = false
    }
    
    if isBalance {
        print("yes")
    } else {
        print("no")
    }
}
